numpy/lib/__init__.py

```diff
@@ -11,51 +11,83 @@
+# `_lib_importer.py` file, which defines `__all__` to include all names to
+# be added to the main namespace!
+
+from . import scimath as emath  # import explicitly to rename
+
 from .arrayterator import Arrayterator
-from .arraypad import *
-from ._version import *
 from numpy.core._multiarray_umath import tracemalloc_domain
-
-__all__ = ['emath', 'math', 'tracemalloc_domain', 'Arrayterator']
-__all__ += type_check.__all__
-__all__ += index_tricks.__all__
-__all__ += function_base.__all__
-__all__ += shape_base.__all__
-__all__ += stride_tricks.__all__
-__all__ += twodim_base.__all__
-__all__ += ufunclike.__all__
-__all__ += arraypad.__all__
-__all__ += polynomial.__all__
-__all__ += utils.__all__
-__all__ += arraysetops.__all__
-__all__ += npyio.__all__
-__all__ += nanfunctions.__all__
-__all__ += histograms.__all__
+from ._version import NumpyVersion
 
 from numpy._pytesttester import PytestTester
 test = PytestTester(__name__)
 del PytestTester
+
+from . import _lib_importer as _hidden_lib_namespace
+
+
+# Ensure that __all__ only contains those things that we want it to contain.
+# Because this is an __init__, we would otherwise include all submodules.
+#
+# NOTE: The symbols not included in `__all__` will be hidden, but do not go
+#       through `__getattr__` (see `__dir__`).
+__all__ = {'emath', 'Arrayterator', 'tracemalloc_domain', 'NumpyVersion'}
+
+__lib_submodules_ = {
+    'type_check',
+    'emath',
+    'arraysetops',
+    'utils',
+    'format',
+    'npyio',
+    'arrayterator',
+    'mixins',
+    'stride_tricks',
+    'twodim_base',
+    'histograms',
+    'index_tricks',
+    'nanfunctions',
+    'shape_base',
+}
+# Add submodules to all (and finalize all)
+__all__.update(__lib_submodules_)
+__all__ = sorted(__all__)
+
+# Add hidden submodules:
+__lib_submodules_.update({
+    'arraypad',
+    'function_base',
+    'polynomial',
+    'scimath',
+    "ufunclike",
+    })
+
+
+def __getattr__(name):
+    if name in __lib_submodules_:
+        import importlib
+        # Allow lazy import (and avoid requiring explicit import), at the
+        # time of writing `mixins` may be the only namespace using this.
+        return importlib.import_module("." + name, __name__)
+
+    try:
+        attribute = getattr(_hidden_lib_namespace, name)
+    except AttributeError:
+        pass
+    else:
+        # Adding a warning in this branch would allow us to properly deprecate
+        # all hidden attributes. Raise an error here to test NumPy itself.
+        return attribute
+
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+
+
+def __dir__():
+    # Overriding __dir__ hides symbols from tab completion for example.
+    # Note that it does _not_ hide symbols from `np.lib.__dict__()`, to do
+    # that, we would have to delete them from the locals here and hide them
+    # into `__getattr__`.  But this seems sufficient, e.g. for tab completion.
+    return __all__
```

Comment:
```suggestion
    return sorted(globals().keys() | set(__all__))
```
More or less equivalent to `sorted(super().__dir__() | set(__all__))` in this case, 
but module-level `__getattr__` lacks a `self` argument.

The import thing here is that you're missing quite a few `types.ModuleType`-based attributes if you just return `__all__` here (and even the `__all__` attribute itself).